# 异常体系及处理

**异常**是程序中的一些错误，但并不是**所有的错误**都是异常，并且错误有时是可以避免的。Java 对异常的处理是按异常分类处理的，不同异常有不同的分类，每种异常都对应一个类型(class)，每个异常都对应一个异常（类的）对象。

* 异常类有两个来源：一是 **Java** 语言本身定义的一些基本异常类型；二是通过继承Exception 类或者其子类自己定义的异常。
* 异常对象的来源：一是 **Java** 运行时环境自动抛出系统生成的异常；二是程序员在处理业务逻辑遇到非预期的情况所抛出的异常。

## 5.1 异常层次结构

![异常层次结构](img/5.1.Ex.png)

**异常层次说明**

* Throwable : 所有异常和错误的超类，它有两个子类 **Error** 和 **Exception** 
* Error：Error是Throwable的子类，表示合理的应用程序不应该尝试捕获的严重问题。 大多数此类错误都是异常情况。 ThreadDeath错误，虽然是“正常”条件，但也是Error的子类，因为大多数应用程序不应该尝试捕获它。一个方法不需要在其throws子句中声明在执行方法期间可能抛出但未捕获的任何Error类，因为这些错误是永远不应发生的异常情况。 也就是说，出于编译时异常检查的目的，Error及其子类被视为未经检查的异常。
* Exception: 类Exception及其子类是Throwable的一种形式，它表示合理的应用程序可能想要捕获的条件。
  类Exception和任何不是RuntimeException子类的子类都是经过检查的异常。 如果方法或构造函数的throws子句可以通过执行方法或构造函数抛出并在方法或构造函数边界外传播，则需要在方法或构造函数的throws子句中声明已检查的异常。
* RuntimeException：这些异常是**Java** 虚拟机正常运行期间抛出的异常的超类。**Java**编译器不去检查它，也就是说当程序中出现这些类异常时，即使没有用**try … catch** 语句捕获它，也没有用**throws**语句声明抛出它，还是会编译通过的，这种异常可以通过改进代码实现来避免。
* 非运行时异常：RuntimeException 意外的异常，类型上属于**Exception**类及其子类。从程序语法角度来讲时必须处理的异常，如果不处理，就编译不通过。这些异常时**Java** 编译器在编译的时候检查的异常。

## 5.2 常见的异常

| 异常类                             | 异常来源                                 |
| ---------------------------------- | ---------------------------------------- |
| **ArithmeticException**            | 算术异常类，比如除0                      |
| **NullPointerException**           | 空指针异常，访问null 对象                |
| **ClassCastException**             | 类型强制转换，类型不匹配的异常           |
| **IndexOutOfBoundsException**      | 越界异常，比如访问数组、链表等超出范围。 |
| **SecurityException**              | 违背安全原则异常                         |
| **ArrayIndexOutOfBoundsException** | 访问数组非法下标异常                     |
| **NoSuchMethodException**          | 方法未找到异常                           |
| **IOException**                    | 输入/输出 异常                           |

[常见异常](https://docs.oracle.com/javase/8/docs/api/)：https://docs.oracle.com/javase/8/docs/api/

## 5.3 异常的捕捉机制

### 5.3.1 异常处理的基本语法

**Java 异常处理范式一**：

```java
try{
  /**执行程序代码块***/
  // 在执行过程中有可能抛异常
}catch(异常类型A 异常变量){
  // 异常处理代码
}catch(异常类型B 异常变量){
  // 异常处理代码
}finally{
  // 异常发生之后，方法返回之前，总要执行的代码
  // 比如：释放锁、关闭文件流。。。。。
  // 事务回滚
}
```

范式一 用于方法返回之前必须要执行最后的收尾代码，比如释放锁、关闭流、事务回滚。为什么收尾代码要在finally 里执行呢？ 是为了防止方法在**某个返回分支**返回时遗留。

**Java 异常处理范式二**

如果方法在返回之前没有必须要执行的收尾代码，可以用范式二。

```java
try{
  /**执行程序代码块***/
  // 在执行过程中有可能抛异常
}catch(异常类型A 异常变量){
  // 异常处理代码
}catch(异常类型B 异常变量){
  
}
```



**Java 异常处理范式三**

范式三，**一般不推荐**，没有异常处理代码块。

```java
try{
  
}finally{
  // 异常发生之后，方法返回之前，总要执行的代码
  // 比如：释放锁、关闭文件流。。。。。
  // 事务回滚
}
```



### 5.3.2 关键字throw、throws

**throw** 关键字用于在一个方法内抛一个异常。如果抛出了检查异常(非RuntimeException)，则还应该在方法头部声明方法可能抛出的异常，则要使用关键字**throws** 声明抛异常。例子：

**抛出RuntimeException**

```java
/**抛出RuntimeException，则方法体不用声明抛出异常**/
	public void testExceptions() {
		throw new RuntimeException("Test runtime exception");
	}
```

**抛出非RuntimeException，则需要声明抛出异常，否则会编译不过去**

```java
 /**在这种情况下，代码编辑器会提示错误，然后右键会提示处理方法：try ... catch 或者 在方法体声明抛异常**/	
 public void testNonRuntimeExceptions(){
		if (Thread.interrupted())  // Clears interrupted status!
		      throw new InterruptedException();
	}
```

```java
	public void testNonRuntimeExceptions() throws InterruptedException{
		if (Thread.interrupted())  // Clears interrupted status!
		      throw new InterruptedException();
	}
```



### 5.3.4 自定义异常

```java
package tianqin.iedu;

public class MyException extends Exception{

	/**
	 * 
	 */
	private static final long serialVersionUID = 2187346212335044701L;
	public MyException() {
		super();
	}
	
	public MyException(String msg) {
		super(msg);
	}
	
}
```



## 5.4 异常最佳实践

**1、只针对异常的情况才使用异常：**

这条规则意味着不要滥用异常，比如用异常控制流程。下面是一个滥用异常的例子：

```java
try{
  int i = 0;
  while(true){
    range[i++].climb();
  }
}catch(Exception e){
   /*****/
}
```

**这种想法的错误**：

* 违背异常机制的设计初衷，异常机制的初衷是用于不正常的情形，所以很少会有JVM实现试图对它们进行优化。
* 把代码放在 **try-catch**块中反而阻止了现代JVM实现本来可能执行的某些特定优化。



**2、对可恢复的情况使用受检异常，对编程错误使用运行时异常**

Java 程序设计语言提供了三种可抛出结构：受检查的异常(**checked exception**)、运行时异常(**run-time exception**)、和错误(**error**)。

主要的原则：

* 如果期望调用者能够适当地恢复，对于这种情况就应该使用受检的异常。受检异常强迫调用者在一个catch子语句中处理该异常，或者将它传播出去。因此，方法中声明要抛出的每个受检的异常，都是对API用户的一种潜在提示：该方法可能会抛出异常，要注意异常的处理。
* 用运行时异常来表示编程错误。大多数的运行时异常都表示前提违例。前提违例是API的调用者没有遵守API规范建立的约定。 比如数组的访问约定指明了下标值必须在[0, length-1]，又比如不能除0。运行时异常是指编程违反了Java 的一些常识（访问数组不能越界、不能除0）。
* Error：一般都是由JVM 系统抛出。



**3、避免不必要地使用受检的异常**

受检的Java 异常与返回代码不同，它们强迫程序员必须处理异常，虽然大大地增强了可靠性，但是**过分使用**则会增加了代码的复杂，所以不要滥用受检异常。

**4、优先使用标准的异常**

​    不在万不得已，谨慎使用自定义异常，一方面代码不通用、另一方面增加了额外的工作。

   建议参考 **5.2 常见的异常**

**5、抛出与抽象相对应的异常**

如果方法抛出的异常与它所执行的任务没有明显的联系，这种情形会使人不知所措。为了避免这个问题，**更高层的实现应该捕获低层的异常，同时抛出可以按照高层抽象进行解释的异常。这种做法被称为异常转译。**

```java
// Exception Translation （Exception Chaining）
try{
  ... // Use lower-level abstraction to do bidding
}catch(LowerLevelException cause){
  throw new HigherLevelException(cause);
}
```



**6、在细节消息中包含能捕获失败的消息**

通过异常的message 将异常原因返回给调用者，比如**IndexOutOfBoundsException**，则如果要抛出异常，建议如下：

```java
/**假如数组默认大小是100，则要进行遍历处理，为了防止异常情况，做了越界判断**/
for(int i = 0; i < 100; i++){
  if(i >= array.length){
    throw IndexOutOfBoundsException(String.format("Out of Bounds: index %d >= lenght %d",     i, array.length));
  }
}
```



**7、努力是失败保持原子性**

一般而言，失败的方法调用应该使对象保持在被调用之前的状态，这叫失败原子性。或者说回滚。

**8、不要忽略异常**

建议有异常要处理，处理异常要处理对。如下的例子就是处理错误的异常：

```java
 void func()throws IOException{
    //......
 }

 void caller(){
   try{
     func();
     // 可能抛出 IOException
   }catch(TimeoutException e){
     // 异常处理: 这段代码逻辑不会处理到
     // 在实际实践中，确实有遇到人会犯类似的错误。
   }
 }
```

一般在处理RuntimeException的时候，这些异常如果API文档没有说明，编程人员也不知道要处理哪些异常，这种情况下，可以这样catch 父类Exception。

```java
void test(){
  try{
    // 业务逻辑，可能抛出异常
  }catch(Exception e){
    // 异常处理
  }
}
```

